pkg/manager/internal.go

```diff
@@ -0,0 +1,178 @@
+	runnables []Runnable
+
+	cache cache.Cache
+
+	// TODO(directxman12): Provide an escape hatch to get individual indexers
+	// client is the client injected into Controllers (and EventHandlers, Sources and Predicates).
+	client client.Client
+
+	// fieldIndexes knows how to add field indexes over the Cache used by this controller,
+	// which can later be consumed via field selectors from the injected client.
+	fieldIndexes client.FieldIndexer
+
+	// recorderProvider is used to generate event recorders that will be injected into Controllers
+	// (and EventHandlers, Sources and Predicates).
+	recorderProvider recorder.Provider
+
+	mu      sync.Mutex
+	started bool
+	errChan chan error
+	stop    <-chan struct{}
+
+	startCache func(stop <-chan struct{}) error
+}
+
+// Add sets dependencies on i, and adds it to the list of runnables to start.
+func (cm *controllerManager) Add(r Runnable) error {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+
+	// Set dependencies on the object
+	if err := cm.SetFields(r); err != nil {
+		return err
+	}
+
+	// Add the runnable to the list
+	cm.runnables = append(cm.runnables, r)
+	if cm.started {
+		// If already started, start the controller
+		go func() {
+			cm.errChan <- r.Start(cm.stop)
+		}()
+	}
+
+	return nil
+}
+
+func (cm *controllerManager) SetFields(i interface{}) error {
+	if _, err := inject.ConfigInto(cm.config, i); err != nil {
+		return err
+	}
+	if _, err := inject.ClientInto(cm.client, i); err != nil {
+		return err
+	}
+	if _, err := inject.SchemeInto(cm.scheme, i); err != nil {
+		return err
+	}
+	if _, err := inject.CacheInto(cm.cache, i); err != nil {
+		return err
+	}
+	if _, err := inject.InjectorInto(cm.SetFields, i); err != nil {
+		return err
+	}
+	if _, err := inject.StopChannelInto(cm.stop, i); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (cm *controllerManager) GetConfig() *rest.Config {
+	return cm.config
+}
+
+func (cm *controllerManager) GetClient() client.Client {
+	return cm.client
+}
+
+func (cm *controllerManager) GetScheme() *runtime.Scheme {
+	return cm.scheme
+}
+
+func (cm *controllerManager) GetFieldIndexer() client.FieldIndexer {
+	return cm.fieldIndexes
+}
+
+func (cm *controllerManager) GetCache() cache.Cache {
+	return cm.cache
+}
+
+func (cm *controllerManager) GetRecorder(name string) record.EventRecorder {
+	return cm.recorderProvider.GetEventRecorderFor(name)
+}
+
+func (cm *controllerManager) Start(stop <-chan struct{}) error {
+	func() {
+		cm.mu.Lock()
+		defer cm.mu.Unlock()
+
+		// Start the Cache.
+		cm.stop = stop
+
+		// Allow the function to start the cache to be mocked out for testing
+		if cm.startCache == nil {
+			cm.startCache = cm.cache.Start
+		}
+		go func() {
+			if err := cm.startCache(stop); err != nil {
+				cm.errChan <- err
+			}
+		}()
+
+		// Wait for the caches to sync.
+		// TODO(community): Check the return value and write a test
+		cm.cache.WaitForCacheSync(stop)
+
+		// Start the runnables after the promises
```

Comment:
comment is outdated (mentions `promises`)

Reply:
Done